### ISSUE TITLE
Bug: Type TSTypeAnnotation not aware of its typeName member

### ISSUE THREAD

**Issue by moniuch** [typescript-eslint/typescript-eslint#6489]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Repro Code

```typescript
constructor(private svc: MyService){}
```

### ESLint Config

```javascript
{
  "parser": "@typescript-eslint/parser"
}
```

### tsconfig

```
 // none, for now compiling and building successfully with `tsup`
```

### Expected Result

Autocomplete suggests `typeAnnotation` member when dotting into the `TSTypeAnnotation`

### Actual Result

Autocomplete doesn't know. Casting to any works and retrieves expected value

### Additional Info

I am developing my first eslint rule, so I might be missing some bits here and there, but I believe I am having problem with `TSESTree.TSTypeAnnotation` type, because autocomplete fails, whereas the value is retrieved properly during runtime.

I have this Angular DI class constructor: `constructor(private svc: MyService){}` where I am trying to get hold of `MyService` identifier.

In AST tree it looks like so:

    {
              "type": "MethodDefinition",
              "key": {
                "type": "Identifier",
                "name": "constructor",
                "range": [
                  142,
                  153
                ]
              },
              "value": {
                "type": "FunctionExpression",
                "id": null,
                "params": [
                  {
                    "type": "TSParameterProperty",
                    "accessibility": "private",
                    "parameter": {
                      "type": "Identifier",
                      "name": "svc",
                      "range": [
                        162,
                        176
                      ],
                      "typeAnnotation": {
                        "type": "TSTypeAnnotation",
                        "range": [
                          165,
                          176
                        ],
                        "typeAnnotation": {
                          "type": "TSTypeReference",
                          "typeName": {
                            "type": "Identifier",
                            "name": "MyService",
                            "range": [
                              167,
                              176
                            ]
                          },
                          "range": [
                            167,
                            176
                          ]
                        }
                      }
                    },
                    "range": [
                      154,
                      176
                    ]
                  }
                ],
                // ...snipped
              },
              "computed": false,
              "static": false,
              "kind": "constructor",
              "override": false,
              "range": [
                142,
                180
              ]
            }

So the structure that leads to `"name": "MyService"` is as follows:

```
parameter [Identifier]
  typeAnnotation [TSTypeAnnotation]  <-- selector's node
    typeAnnotation [TSTypeReference]
      typeName [Identifier]
        name: MyService
```

Since in my rule I am selecting `typeAnnotation [TSTypeAnnotation]`

```ts
    const selectors = {
      ctorParam: [
        AST_NODE_TYPES.ClassBody,
        `${AST_NODE_TYPES.MethodDefinition}[kind="constructor"]`,
        AST_NODE_TYPES.FunctionExpression,
        AST_NODE_TYPES.TSParameterProperty,
        AST_NODE_TYPES.Identifier,
        AST_NODE_TYPES.TSTypeAnnotation,
      ].join(' ')
    }
```

```ts
    [selectors.ctorParam](node: TSESTree.TSTypeAnnotation) {
      node.typeAnnotation.typeName.name // typeName gets TS squiggles
      // console.log((node.typeAnnotation as any).typeName.name) <-- this works, outputs the expected value
    },
```

I would expect to see `typeName` among the autocomplete suggestions (please scroll down to the bottom of this ticket for the screenshot).

```
### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `n/a` |
| `@typescript-eslint/parser`            | `5.52.0` |
| `@typescript-eslint/scope-manager`     | `5.52.0` |
| `@typescript-eslint/typescript-estree` | `5.52.0` |
| `@typescript-eslint/type-utils`        | `n/a` | 
| `@typescript-eslint/utils`             | `5.52.0` |
| `TypeScript`                           | `4.9.5` |
| `ESLint`                               | `8.34.0` |
| `node`                                 | `18.10.0` |


* `@typescript-eslint/type-utils` - I had been running without it. Installed it manually just to see if it helps, even though _This is an internal package to the typescript-eslint monorepo. You likely don't want to use it directly._  It didn't help. Gonna uninstall it.
```

Autocomplete

![image](https://user-images.githubusercontent.com/681956/219899671-52b26339-91ba-40b6-b8bd-f49edcb7e014.png)

**Comment by armano2** [comment]
this is workign as expected, [`typeAnnotation`](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/ast-spec/src/special/TSTypeAnnotation/spec.ts#L7) can be many things
https://github.com/typescript-eslint/typescript-eslint/blob/16144d1d6a37668393f731913a306008f33eeac6/packages/ast-spec/src/unions/TypeNode.ts#L47-L92

you want to verify if `type` of `typeAnnotation` node is `TSTypeReference`
```ts
if (node.typeAnnotation.type === AST_NODE_TYPES.TSTypeReference) {
   node.typeAnnotation.typeName
}
```

![image](https://user-images.githubusercontent.com/625469/219901741-88aabc40-7770-4b2e-80f8-28b31e98a670.png)

**Comment by moniuch** [comment]
@armano2 Thank you. Now I see: a discriminated union! I have noticed a similar problem with import specifiers and realized I must be doing something wrong :)
